cyberbrain/value_stack.py

```diff
@@ -88,16 +84,98 @@ class Why(enum.Enum):
+        start_lineno: The starting lineno of the item.
+        sources: A list of symbols that are the sources of the item.
+    """
+
+    def __init__(
+        self,
+        start_lineno: int,
+        sources: List[Symbol],
+    ):
+        self.start_lineno = start_lineno
+        self.sources = sources
+
+    def __repr__(self) -> str:
+        return "{" + repr(self.start_lineno) + " " + repr(self.sources) + "}"
+
+    def get_top_source(self) -> Symbol:
+        # This happens when variables (e.g. globals) that we don't care about are mutated.
+        if len(self.sources) == 0:
+            return None
+        return self.sources[0]
+
+    def copy(self) -> SymbolStackItem:
+        return SymbolStackItem(self.start_lineno, [copy(sym) for sym in self.sources])
+
+
+class CustomValueStackItem:
+    """Class representing a custom value (Exceptions, Why's, anything not tracked by Cyberbrain) on the value stack
+
+    Attributes:
+        custom_value: The value of the custom item.
+    """
+
+    def __init__(
+        self,
+        custom_value: any = None,
+    ):
+        self.custom_value = custom_value
+
+    def __repr__(self) -> str:
+        return repr(self.custom_value)
+
+    def copy(self) -> CustomValueStackItem:
+        return CustomValueStackItem(self.custom_value)
+
+
+class SymbolWithCustomValueStackItem(SymbolStackItem, CustomValueStackItem):
+    """Class representing a both a symbol and a custom value on the value stack
+
+    Some types like None's and Exceptions can be pushed onto the stack as either
+    a normal SymbolStackItem or a CustomValueStackItem used for exceptions.
+    Therefore, this class inherits from both SymbolStackItem and CustomValueStackItem for flexibility.
+    """
+
+    def __init__(self, start_lineno: int, sources: List[Symbol], custom_value=None):
+        SymbolStackItem.__init__(self, start_lineno, sources)
+        CustomValueStackItem.__init__(self, custom_value)
+
+    def copy(self) -> SymbolWithCustomValueStackItem:
+        return SymbolWithCustomValueStackItem(
+            self.start_lineno, self.sources, self.custom_value
+        )
+
+    def __repr__(self) -> str:
+        return "SymbolWithCustomValueStackItem{" + repr(self.custom_value) + "}"
+
+
+# Placeholder for the items on the stack that we don't care about.
+# This could be from a LOAD_CONST but since we can get the value later,
+# we don't care about the value pushed on the stack
+_placeholder = SymbolStackItem(-1, [])
+
+
+# Placeholder for a None on the stack
+# Could be a None that is stored in a variable or a None used by exception handling.
+_none_placeholder = SymbolWithCustomValueStackItem(-1, [], None)
+
+
 class BaseValueStack:
     """Class that simulates the a frame's value stack.
 
     This class contains instr handlers that are the same across different versions.
     """
 
     def __init__(self):
-        self.stack = []
+        self.stack: List[SymbolStackItem | CustomValueStackItem] = []
```

Comment:
We can't use this because we need to support older versions
https://www.python.org/dev/peps/pep-0604/ is added in 3.10

Reply:
Ditto for other usages of `|`

Reply:
Got it